src/Storages/MergeTree/MergeTreeData.cpp

```diff
@@ -3948,68 +3941,73 @@ void MergeTreeData::forcefullyMovePartToDetachedAndRemoveFromMemory(const MergeT
             restored_active_part = true;
         };
 
-        auto it_middle = data_parts_by_info.lower_bound(part->info);
-
-        /// Restore the leftmost part covered by the part
-        if (it_middle != data_parts_by_info.begin())
+        /// ActiveDataPartSet allows to restore most top-level parts instead of unexpected.
+        /// It can be important in case of assigned merges. If unexpected part is result of some
+        /// finished, but not committed merge we should restore closest ancestors for the
+        /// unexpected part to be able to execute it.
```

Comment:
It's not guaranteed to restore the most top-level parts in case of intersecting parts (it will choose _some_ parts). But it's okay. Also, maybe we should sort parts by level (descending) before adding them to `parts_for_replacement` set (so bigger parts will have priority)